src/BuiltInTools/DotNetDeltaApplier/HotReloadAgent.cs

```diff
@@ -45,116 +44,211 @@ private void OnAssemblyLoad(object? _, AssemblyLoadEventArgs eventArgs)
-                        {
-                            try
-                            {
-                                action(types);
-                            }
-                            catch (Exception ex)
-                            {
-                                _log($"Exception from '{action}': {ex}");
-                            }
-                        };
+                        Visit(assemblies, dependency, sortedAssemblies, visited);
                     }
+                }
+
+                sortedAssemblies.Add(assembly);
+            }
 
-                    MethodInfo? GetUpdateMethod(Type handlerType, string name)
+            static Assembly? FindDependency(ReadOnlySpan<Assembly> assemblies, AssemblyName dependencyName)
+            {
+                foreach (var assembly in assemblies)
+                {
+                    if (assembly.FullName == dependencyName.FullName)
                     {
-                        if (handlerType.GetMethod(name, BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static, new[] { typeof(Type[]) }) is MethodInfo updateMethod &&
-                            updateMethod.ReturnType == typeof(void))
-                        {
-                            return updateMethod;
-                        }
-
-                        foreach (MethodInfo method in handlerType.GetMethods(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static | BindingFlags.Instance))
-                        {
-                            if (method.Name == name)
-                            {
-                                _log($"Type '{handlerType}' has method '{method}' that does not match the required signature.");
-                                break;
-                            }
-                        }
-
-                        return null;
+                        return assembly;
                     }
                 }
+
+                return null;
             }
 
-            return new UpdateHandlerActions(before, after);
+            return sortedAssemblies;
         }
 
         public void ApplyDeltas(IReadOnlyList<UpdateDelta> deltas)
         {
             try
             {
-                UpdateHandlerActions beforeAfterUpdates = _beforeAfterUpdates ??= GetMetadataUpdateHandlerActions();
+                // Defer discovering the receiving deltas until the first hot reload delta.
+                // This should give enough opportunity for AppDomain.GetAssemblies() to be sufficiently populated.
+                _handlerActions ??= GetMetadataUpdateHandlerActions();
+                var handlerActions = _handlerActions;
+
+                // TODO: Get types to pass in
+                Type[]? updatedTypes = null;
 
-                beforeAfterUpdates.Before.ForEach(b => b(null)); // TODO: Get types to pass in
+                InvokeActions(handlerActions.Before, updatedTypes);
 
-                foreach (var item in deltas)
```

Comment:
What was wrong with `foreach` ?

Reply:
I had it updated to use a local and I was worried I would get complaints about capturing delegates. I reverted it - the original version had less code 👍🏽